listenbrainz/webserver/views/test/test_user.py

```diff
@@ -131,7 +131,7 @@ def test_username_case(self):
         response2 = self.client.get(url_for('user.profile', user_name='IlieKcomPUteRs'))
         self.assert200(response1)
         self.assert200(response2)
-        self.assertEqual(response1.data.decode('utf-8'), response2.data.decode('utf-8'))
+        self.assertEqual(response1.data.decode('utf-8')[:100], response2.data.decode('utf-8')[:100])
```

Comment:
I wonder if we can be smarter here. 2 things come to mind:
 - We probably shouldn't enable the debug toolbar during testing - we probably want to be closer to "production" here rather than development. I don't know if there's a way to disable DEBUG during tests, or explicitly disable the debug toolbar. I'll look into it
- An alternative to checking the whole response page is to just check the user that we pass to the template. I know that in django you can check the template context in a test: https://docs.djangoproject.com/en/2.1/topics/testing/tools/#django.test.Response.context
Let's see if we can do the same in flask

Reply:
Turning off debugging during tests sounds great, but I wonder why this error didn't occur during the previous version of flask? I'm not sure why it didn't run into this issue previously.

Reply:
Looks like flask also has a `render_template()` function :)

Reply:
It might be worth just doing:
```python
        self.assertTrue("iliekcomputers" in response1.data.decode('utf-8'))
        self.assertTrue("iliekcomputers" in response2.data.decode('utf-8'))
```
For readability & efficiency?
I imagine that searching for that exact string would be far faster than rendering the template to check it.

Reply:
There is a `self.assertIn` in Flask-Testing

Reply:
The django method doesn't test the rendered html - instead it gives you the actual parameter that was passed to the template ('user' in this case). This test is really only that you get the correct username even if the capitalisation is different, so ideally we should only check this user object.

Reply:
Ohh okay that makes more sense 👍

Reply:
https://github.com/metabrainz/listenbrainz-server/blob/60f66aea236901ae225b7b495d4da0afc9fb9fb7/listenbrainz/webserver/views/test/test_user.py#L58

Reply:
I think we're talking about `self.assertContext`

Reply:
perfect, that's exactly what I mean! I didn't know what it was called in flask. let's replace this assertEqual with an assertContext for the 'user' parameter.
@Ge0rg3 has a good point though - it's interesting to know why this didn't happen when we were using older versions of flask. I know they made some changes to the way that debug mode works some time around flask 0.11 or 0.12, but I'm not sure exactly what these changes were

Reply:
That looks good, but I'm currently getting:
```
AssertionError: Context variable does not exist: user.musicbrainz_id
```
From what I can tell from the [template](https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz/webserver/templates/user/profile.html), "user" is an object that contains the ID that we want.
So it doesn't look like assertContext can do properties. Might be worth [getting the context variable itself](https://github.com/jarus/flask-testing/blob/master/flask_testing/utils.py#L253) and checking the musicbrainz_id from there.

Reply:
Okay--I've got it working as follows:
```python
self.assertEqual(self.get_context_variable('user').musicbrainz_id, "iliekcomputers")
```
Am I okay to commit that, or is there any way I can improve it? :)

Reply:
Is there a reason why `self.assertContext('user', self.user)` doesn't work? We should be able to directly check these objects instead of accessing fields within them.

Reply:
From what I can tell from the [assertContext source code](https://github.com/jarus/flask-testing/blob/master/flask_testing/utils.py#L283), it can't read the (sub?)attributes of user, since we need to read the `musicbrainz_id` attribute of `user`. It also uses the get_context_variable function itself.

Reply:
I see what you're trying to do. I don't think that we need to specifically compare the `musicbrainz_id` attribute. Instead, we can directly compare the user object, which uses `flask_login`'s UserMixin and its equals method (to compare user id) https://flask-login.readthedocs.io/en/latest/_modules/flask_login/mixins.html#UserMixin

Reply:
(See edited comment above)

Reply:
you're right that the definition of "equals" could be different. We are lucky because our User object is a flask_login UserMixin. This allows us to use the definition of `__eq__` which checks `self.id`. If we were not using this, we could have an issue where the two objects that we wanted to test might have been different.

Reply:
You should make sure that you add 2 `self.asserContext`s. One for each `self.client.get`. These checks should be directly after the get - because the context changes after each one

Reply:
Ah okay, will do :)

Reply:
Interesting, I'm getting this error when running the tests:
```

    self.assertEqual(self.get_context_variable(name), value)
E   AssertionError: <listenbrainz.webserver.login.User object[14 chars]cfd0> != (1, datetime.datetime(2018, 11, 6, 11, 14[348 chars]e, 1)
```

When at the line:
```
        response1 = self.client.get(url_for('user.profile', user_name='iliekcomputers'))
>       self.assertContext('user', self.user)
```

Reply:
So at that stage, self.user is equal to:
```
(1, datetime.datetime(2018, 11, 6, 11, 31, 17, 726742, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'iliekcomputers', '5b879824-9591-48dd-82e6-268400d77633', datetime.datetime(2018, 11, 6, 11, 31, 17, 726742, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), datetime.datetime(1970, 1, 1, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), None, 1)
```

Reply:
> we could have an issue where the two objects that we wanted to test might have been different.

I think that this might actually be the case :/ @alastair

Reply:
Thanks for the info. I'll get to work on that 👍

Reply:
@alastair I've implemented these channges, and added them in the new commit.
Just for reference, I used `get_or_create` instead of just `create` so that the user is only created if it doesn't exist (and it saves me doing a database search for it).